### ISSUE TITLE
[go_router_builder] GoRouteData `redirect` function has no context nor state parameters

### ISSUE THREAD

**Issue by lucavenir** [flutter/flutter#115955]
## TLDR
The pain point described here is that `GoRouteData`'s `redirect` differs, or rather lacks, from `GoRouter`'s. See [GoRouter's redirect](https://pub.dev/documentation/go_router/latest/go_router/GoRouterRedirect.html) vs [GoRouteData's redirect](https://pub.dev/documentation/go_router/latest/go_router/GoRouteData/redirect.html).

Without `GoRouteData` matching `GoRouter` 's `redirect` APIs, I am unable to inject state or to read the request sublocation and therefore I have to give up on typed routing.

The root of the problem arises from integrating `GoRouter` with `Riverpod` (or `Provider` for what matters).

## Use case

I am venturing through `go_router_builder` and I am trying to make it work with my state management solution.
So far, so good.

Then, I tried implementing a sub-route with a route-level redirection. Say you have the following (simplified) routes:
```
+ splash
+ login
+ home
++ admin
++ dashboard
++ guest-view
```

<details>
<summary>Details of the use case</summary>

> The user opens the app and sees a splash page. He is not logged in, and therefore should be redirected towards `login`. Then, the users chooses to login as a guest. Just like that, he's redirected to `home`. The user tries to navigate to `admin`, but he's redirected back to `home` since he has not enough permissions. Some sort of feedback is showed to the user.

</details>

As detailed above, the crucial part here is being able to write route-level logic that redirects an _unauthenticated but not unauthorized_ user, e.g. a guest user trying to access an admin page. Without codegen, I usually do this via [route-level redirects](https://pub.dev/documentation/go_router/latest/topics/Redirection-topic.html). I read some state via `ref.read` (Riverpod) and then make decisions based on those results.

Unluckily, this is not currently possible via `GoRouteData` and the codegen APIs. Or at least [it looks like it](https://pub.dev/packages/go_router_builder#route-level-redirection).

## Proposal

Let `GoRouteData` have a `FutureOr<String?> Function(BuildContext, GoRouterState) redirect` parameter instead of the current simple `FutureOr<String?> redirect()`.

This allows the developer to inject state through `BuildContext` and use `GoRouterState` when needed.

**Comment by exaby73** [comment]
Hello @lucavenir. Thank you for the proposal. Looks like a valid one to me. Labeling for visibility by the team

**Comment by lucavenir** [comment]
Thank you @exaby73. I want to underline how this behavior might be desirable with `builder` in `GoRouteData`. Note how that method lacks a `GoRouterState` parameter (which is normally found in `GoRouter`).

That _might_ be de desirable, I guess that works, as we're defining `GoRouteData` to _directly_ accept typed parameters in its subclass, and let `go_router_builder` generate those APIs for us.

But in case of `redirect` I don't think it's possible to write logic without a `BuildContext` (i.e. _global state_) or on `GoRouterState` (sublocation requested by the user).

If there's a way to make that work, please let me know, i.e. this becomes a documentation request rather than a feature request.
